libvips/foreign/webpsave.c

```diff
@@ -109,15 +155,639 @@ typedef struct _VipsForeignSaveWebp {
+
+	/* Add metadata with this.
+	 */
+	WebPMux *mux;
+
+	/* The current frame coming from libvips, and the y position
+	 * in the input image.
+	 */
+	VipsRegion *frame;
+	int write_y;
 
-} VipsForeignSaveWebp;
+	/* VipsRegion is not always contiguious, but we need contiguous RGB(A)
+	 * for libwebp. We need to copy each frame to a local buffer.
+	 */
+	VipsPel *frame_bytes;
+} VipsForeignSaveWebP;
 
-typedef VipsForeignSaveClass VipsForeignSaveWebpClass;
+typedef VipsForeignSaveClass VipsForeignSaveWebPClass;
 
-G_DEFINE_ABSTRACT_TYPE( VipsForeignSaveWebp, vips_foreign_save_webp, 
+G_DEFINE_ABSTRACT_TYPE( VipsForeignSaveWebP, vips_foreign_save_webp,
 	VIPS_TYPE_FOREIGN_SAVE );
 
+static void
+vips_webp_write_unset( VipsForeignSaveWebP *write )
+{
+	WebPMemoryWriterClear( &write->memory_writer );
+	VIPS_FREEF( WebPAnimEncoderDelete, write->enc );
+	VIPS_FREEF( WebPMuxDelete, write->mux );
+	VIPS_UNREF( write->image );
+}
+
+static void
+vips_foreign_save_webp_dispose( GObject *gobject )
+{
+	VipsForeignSaveWebP *webp= (VipsForeignSaveWebP *) gobject;
+
+	VIPS_UNREF( webp->frame );
+
+	VIPS_UNREF( webp->target );
+
+	VIPS_FREE( webp->frame_bytes );
+
+	G_OBJECT_CLASS( vips_foreign_save_webp_parent_class )->
+		dispose( gobject );
+}
+
+static gboolean
+vips_webp_pic_init( VipsForeignSaveWebP *write, WebPPicture *pic )
+{
+	if( !WebPPictureInit( pic ) ) {
+		vips_error( "webpsave", "%s", _( "picture version error" ) );
+		return( FALSE );
+	}
+	pic->writer = WebPMemoryWrite;
+	pic->custom_ptr = (void *) &write->memory_writer;
+
+	/* Smart subsampling needs use_argb because it is applied during
+	 * RGB to YUV conversion.
+	 */
+	pic->use_argb = write->lossless ||
+		write->near_lossless ||
+		write->smart_subsample;
+
+	return( TRUE );
+}
+
+/* Write a VipsImage into an unintialised pic.
+ */
+static int
+write_webp_image( VipsForeignSaveWebP *write, const VipsPel *imagedata, WebPPicture *pic )
+{
+	webp_import import;
+	int page_height = vips_image_get_page_height( write->image );
+
+	if( !vips_webp_pic_init( write, pic ) )
+		return( -1 );
+
+	pic->width = write->image->Xsize;
+	pic->height = page_height;
+
+	if( write->image->Bands == 4 )
+		import = WebPPictureImportRGBA;
+	else
+		import = WebPPictureImportRGB;
+
+	if( !import( pic, imagedata,
+		write->image->Xsize * write->image->Bands ) ) {
+		WebPPictureFree( pic );
+		vips_error( "webpsave", "%s", _( "picture memory error" ) );
+		return( -1 );
+	}
+
+	return( 0 );
+}
+
+/* We have a complete frame --- write!
+ */
+static int
+vips_foreign_save_webp_write_frame( VipsForeignSaveWebP *webp)
+{
+	WebPPicture pic;
+	int y;
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( webp );
+	VipsRect *frame_rect = &webp->frame->valid;
+	int page_index = frame_rect->top / frame_rect->height;
+
+	/* We need the frame as a contiguous RGB(A) buffer for libwebp.
+	 */
+	for( y = 0; y < frame_rect->height; y++ )
+		memcpy( webp->frame_bytes + y * webp->image->Bands * frame_rect->width,
```

Comment:
This needs to fit in 80 columns. Perhaps add a pointer that steps down the output to get rid of this huge expression?